--- a/scripts/dtc/Makefile
+++ b/scripts/dtc/Makefile
@@ -23,8 +23,9 @@ HOSTLDLIBS_dtc	:= -lyaml
 endif

 # Generated files need one more search path to include headers in source tree
-HOSTCFLAGS_dtc-lexer.lex.o := -I$(src)
-HOSTCFLAGS_dtc-parser.tab.o := -I$(src)
+HOSTCFLAGS_dtc-lexer.lex.o := -I$(src) -fcommon
+HOSTCFLAGS_dtc-parser.tab.o := -I$(src) -fcommon
+

 # dependencies on generated files need to be listed explicitly
 $(obj)/dtc-lexer.lex.o: $(obj)/dtc-parser.tab.h
